### PR TITLE
fix transfer stops

### PR DESCRIPTION
**Pull request by bcarlin**

### CODE REVIEW

File: src/main/java/org/waarp/openr66/protocol/networkhandler/ssl/NetworkSslServerInitializer.java [hunk 1]
```diff
@@ -70,24 +70,24 @@ protected void initChannel(SocketChannel ch) throws Exception {
...
+        pipeline.addLast(NetworkServerInitializer.LIMITGLOBAL,
+                Configuration.configuration.getGlobalTrafficShapingHandler());
+        // Per channel limitation
+        pipeline.addLast(NetworkServerInitializer.LIMITCHANNEL,
+                new ChannelTrafficShapingHandler(
+                    Configuration.configuration.getServerChannelWriteLimit(),
+                    Configuration.configuration.getServerChannelReadLimit(),
+                    Configuration.configuration.getDelayLimit()));
+ 
+        pipeline.addLast("codec", new NetworkPacketCodec());
```
Comment (fredericBregier) [inline src/main/java/org/waarp/openr66/protocol/networkhandler/ssl/NetworkSslServerInitializer.java:94]: Does this should not be before any TrafficShaping ? not so sure on my side... Maybe it is the right position there...

File: src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerInitializer.java [hunk 2]
```diff
@@ -49,23 +49,21 @@ public NetworkServerInitializer(boolean server) {
...
+        // Global limitation
+        pipeline.addLast(LIMITGLOBAL,
+                Configuration.configuration.getGlobalTrafficShapingHandler());
+        // Per channel limitation
+        pipeline.addLast(LIMITCHANNEL,
+                new ChannelTrafficShapingHandler(
+                    Configuration.configuration.getServerChannelWriteLimit(),
+                    Configuration.configuration.getServerChannelReadLimit(),
+                    Configuration.configuration.getDelayLimit()));
         pipeline.addLast("codec", new NetworkPacketCodec());
```
Comment (fredericBregier) [inline src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerInitializer.java:73]: Same here: before the codec was set previously to traffic shaping... Maybe this is the right place there ?

File: src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerInitializer.java [hunk 3]
```diff
@@ -49,23 +49,21 @@ public NetworkServerInitializer(boolean server) {
     @Override
     protected void initChannel(SocketChannel ch) throws Exception {
         final ChannelPipeline pipeline = ch.pipeline();
+        pipeline.addLast(TIMEOUT, new IdleStateHandler(0, 0,
+                    Configuration.configuration.getTIMEOUTCON(),
+                    TimeUnit.MILLISECONDS));
+        // Global limitation
+        pipeline.addLast(LIMITGLOBAL,
+                Configuration.configuration.getGlobalTrafficShapingHandler());
```
Comment (fredericBregier) [inline src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerInitializer.java]: This never returns null ?

File: src/main/java/org/waarp/openr66/protocol/networkhandler/ssl/NetworkSslServerInitializer.java [hunk 4]
```diff
@@ -70,24 +70,24 @@ protected void initChannel(SocketChannel ch) throws Exception {
...
-        }
-        pipeline.addLast(Configuration.configuration.getHandlerGroup(), "handler", new NetworkSslServerHandler(
-                !this.isClient));
+                new IdleStateHandler(0, 0,
+                    Configuration.configuration.getTIMEOUTCON(),
+                    TimeUnit.MILLISECONDS));
+
+        // Global limitation
+        pipeline.addLast(NetworkServerInitializer.LIMITGLOBAL,
+                Configuration.configuration.getGlobalTrafficShapingHandler());
```
Comment (fredericBregier) [inline src/main/java/org/waarp/openr66/protocol/networkhandler/ssl/NetworkSslServerInitializer.java]: Same: never null ?

File: src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerHandler.java [hunk 5]
```diff
@@ -274,7 +274,8 @@ public void channelRead0(ChannelHandlerContext ctx, NetworkPacket msg) throws Ex
             packet.clear();
             return;
         }
-        logger.debug("GET MSG: " + packet.getCode());
+        logger.debug("GET MSG: {}" + packet.getCode());
+        logger.trace("MSG: {}" + packet);
```
Comment (fredericBregier) [inline src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerHandler.java]: if using `{}`, then you should not use `+` but a comma, right ?
    `logger.trace("MSG: {}", packet);`

File: src/main/java/org/waarp/openr66/protocol/localhandler/TransferActions.java [hunk 6]
```diff
@@ -483,6 +484,22 @@ public void request(LocalChannel channel, RequestPacket packet)
...
+        logger.trace("Received limit {}", packet.getLimit());
+        logger.trace("Local limit {}", localChannelReference.getChannelLimit(runner.isSender()));
+        if (localLimit <= 0) {
+            localLimit = remoteLimit;
+        } else if (remoteLimit > 0 && remoteLimit < localLimit) {
+            localLimit = remoteLimit;
+        }
+        localChannelReference.setChannelLimit(runner.isSender(), localLimit);
+        packet.setLimit(localLimit);
+
```
Comment (fredericBregier) [inline src/main/java/org/waarp/openr66/protocol/localhandler/TransferActions.java:502]: Just a quick question: did you manage a client in an older version, such that limit is not sent, and therefore not in the RequestPacket ? What is the behavior (on both side, even, server older than client) ?

File: src/main/java/org/waarp/openr66/protocol/localhandler/LocalTransaction.java [hunk 7]
```diff
@@ -314,20 +313,6 @@ public LocalChannelReference createNewClient(NetworkChannelReference networkChan
...
-            if (lcr == null) {
-                try {
-                    Thread.sleep(Configuration.RETRYINMS);
-                    Thread.yield();
-                } catch (InterruptedException e) {
-                }
-            } else {
-                return lcr;
-            }
-        }
```
Comment (fredericBregier) [inline src/main/java/org/waarp/openr66/protocol/localhandler/LocalTransaction.java:330]: I remember this code, when I noticed that sometime, there were a delay, without any explanation...
If there is no more issue, then you're right!!

**Review by fredericBregier — COMMENTED** [review]
Some small checks on your side, probably nothing wrong. Just to be sure ;-)

File: src/main/java/org/waarp/openr66/protocol/localhandler/TransferActions.java (hunk 6 above)
Comment (marakiis) [reply src/main/java/org/waarp/openr66/protocol/localhandler/TransferActions.java:502]: The Limit field is carried on the Json part of the RequestPacket.
Older R66 partner not using Json Formating of the packet are not concerned by this change.
Older R66 partner using Json will not parse the Limit field and send a limitless RequestPacket.

(The Json formating of the RequestPacket make it the easiest place to implement changes without breaking retro-compatibility)

File: src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerInitializer.java (hunk 3 above)
Comment (marakiis) [reply src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerInitializer.java]: In the current implementation this shouldn't return null.
But maybe a better solution would be to implement `getGlobalTrafficShapingHandler` as a "singleton" getter:
```
public synchronized GlobalTrafficStapingHandler getGlobalTrafficShapingHandler() {
    if (globalTrafficShapingHandler == null) {
        globalTrafficShapingHandler = new GlobalTrafficShapingHandler(x, y, z);
    }
    return globalTrafficShapingHandler;
}
```

File: src/main/java/org/waarp/openr66/protocol/networkhandler/ssl/NetworkSslServerInitializer.java (hunk 4 above)
Comment (marakiis) [reply src/main/java/org/waarp/openr66/protocol/networkhandler/ssl/NetworkSslServerInitializer.java]: See above

File: src/main/java/org/waarp/openr66/protocol/networkhandler/ssl/NetworkSslServerInitializer.java (hunk 1 above)
Comment (marakiis) [reply src/main/java/org/waarp/openr66/protocol/networkhandler/ssl/NetworkSslServerInitializer.java:94]: Putting the TrafficShapinghandlers before the Codec seems a better way to organize the channel.

The previous configuration limited the Codec speed and not the network writing/reading speed. The Codec was fed bytes as the pace imposed by the TrafficShapinghandlers but was writing on the network as fast as the connection allowed it.

The previous pipeline forced the use of custom TrafficShapinghandlers to reflect the proper size of the network traffic. (It just added 13 bytes to the traffic count to take account of the overhead added by the NetworkCodec).

This pipeline  configuration is cleaner on the traffic shaping, the traffic computation, and allows the use standard netty ShapingHandler.

File: src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerInitializer.java (hunk 2 above)
Comment (marakiis) [reply src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerInitializer.java:73]: See above

File: src/main/java/org/waarp/openr66/protocol/networkhandler/ssl/NetworkSslServerInitializer.java (hunk 1 above)
Comment (fredericBregier) [reply src/main/java/org/waarp/openr66/protocol/networkhandler/ssl/NetworkSslServerInitializer.java:94]: perfect

File: src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerInitializer.java (hunk 3 above)
Comment (fredericBregier) [reply src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerInitializer.java]: hum, `synchronized` will sync all threads trying to connect at the same time, which could be not the correct behavior... (less efficiency in multithtread environment)
so make sure the `getGlobalTrafficShapingHandler()` never returns null (initialization should not allow null traffic handler

File: src/main/java/org/waarp/openr66/protocol/localhandler/TransferActions.java (hunk 6 above)
Comment (fredericBregier) [reply src/main/java/org/waarp/openr66/protocol/localhandler/TransferActions.java:502]: Yes, I know the parameter will be either absent or ignored, according to the version, but my question was more on the "procedure" part: is this taking into account such parameter being absent (so 0) on client or server side being the new version (and therefore taking into consideration this _missing_ parameter) ?

File: src/main/java/org/waarp/openr66/protocol/localhandler/LocalTransaction.java (hunk 7 above)
Comment (bcarlin) [reply src/main/java/org/waarp/openr66/protocol/localhandler/LocalTransaction.java:330]: I tried every possible scenarii, and removing this loop does not present any issue.
However, in this case, when a transfer is stopped, it delays the processing of each packet by 1s from the moment that lcr is closed. It does no harm for a single transfer, but when several transfers between with the same server are stopped at the same time, the total processing delay can extend to several minutes... hence the removal!

File: src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerHandler.java (hunk 5 above)
Comment (bcarlin) [reply src/main/java/org/waarp/openr66/protocol/networkhandler/NetworkServerHandler.java]: fixed in be173d7

File: src/main/java/org/waarp/openr66/protocol/localhandler/TransferActions.java (hunk 6 above)
Comment (bcarlin) [reply src/main/java/org/waarp/openr66/protocol/localhandler/TransferActions.java:502]: Test have been made between old version and this one  : it works well for get and put.

**Pull request merged**